Ability to use local Docker compose binary instead of container.

outofcoffee opened the pull request:
## Purpose
Adds the ability to use local Docker compose binary instead of container.

## Rationale
In scenarios where tests are running within a container (such as using [Jenkins Docker Plugin](https://wiki.jenkins-ci.org/display/JENKINS/Docker+Plugin)), it is difficult for the `docker-compose.yml` file to be bind mounted into the internal Docker compose container used by the current `DockerComposeContainer` implementation.

## Implementation
Key methods from `DockerCompose` have been extracted to an interface, and a second implementation, `LocalDockerCompose` has been added that uses the `docker-compose` on the `PATH`.

Users call the `withLocalCompose()` method on `DockerComposeContainer` to activate this behaviour.

rnorth commented on core/src/main/java/org/testcontainers/containers/DockerComposeContainer.java:
```diff
@@ -305,4 +334,76 @@ public void start() {
+            final Process p = pb.start();
+
+            // wait for the compose process to stop, which should only happen after it has spawned all the service containers
+            p.waitFor(120, TimeUnit.SECONDS);
```
Should this timeout be longer, especially to allow for pulling/building of containers..? 
I'm not sure there's actually any easy timeout length, so maybe just let it be infinite?

rnorth commented on core/src/main/java/org/testcontainers/containers/DockerComposeContainer.java:
```diff
@@ -305,4 +334,76 @@ public void start() {
+                }
+            }
+            
+            if (0 != p.exitValue()) {
```
`p` can't be null, so is it necessary to write like this?

rnorth commented on core/src/main/java/org/testcontainers/containers/DockerComposeContainer.java:
```diff
@@ -305,4 +334,76 @@ public void start() {
+    @Override
+    public void start() {
+        try {
+            final ProcessBuilder pb = new ProcessBuilder(Splitter.onPattern(" ")
```
ah - please use [Zeroturnaround Exec](https://github.com/zeroturnaround/zt-exec) (already on the class path!) for this. It will make a lot of things easier, including pumping stdout/stderr to the logger.

rnorth commented on core/src/main/java/org/testcontainers/containers/DockerComposeContainer.java:
```diff
@@ -305,4 +334,76 @@ public void start() {
+
+    @Override
+    public void start() {
+        try {
```
Please can you use [CommandLine#executableExists](https://github.com/testcontainers/testcontainers-java/blob/master/core/src/main/java/org/testcontainers/utility/CommandLine.java#L58) to check that docker-compose is available on the path first? That way we can fail with an informative error message

rnorth replied:
This comment relates to the `start()` method, BTW

qoomon commented:
I think it shouldn't be done with an explicit call. I would prefer to do it with an environment variable. We than can run the same java source within different environments without adopting the test to specific environment.

rnorth commented:
@qoomon good point. How about we make [`withLocalCompose()`](https://github.com/testcontainers/testcontainers-java/pull/200/files#diff-f992b47914d235d916c506932b1946f6R286) require a boolean value to be passed to turn this behaviour on/off?

That way users would be free to define an environment variable (rather than us dictating one) and use that. Alternatively they could just pass in true if there's something about their usage that means this behaviour is always desired. What do you think @outofcoffee ?

qoomon commented:
I dont like the idea to check out a project and then are not be able to run all test because of environment dependencies. Therefore i would like to keep dependencies at a bare minimum.
Maybe another solution could be to fall back to compose container if withLocalComose() is set.

qoomon commented:
And i think the problem described is already solved by #189

rnorth commented:
What do you think of #189 @outofcoffee? Does this PR address a different
issue?
On Mon, 15 Aug 2016 at 23:17 Bengt Brodersen <notifications@github.com>
wrote:

> And i think the problem described is already solved by #189
> <https://github.com/testcontainers/testcontainers-java/pull/189>
>
> —
> You are receiving this because you commented.
>
>
> Reply to this email directly, view it on GitHub
> <https://github.com/testcontainers/testcontainers-java/pull/200#issuecomment-239946397>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAIET9Nrss9l6RNup_0tHn0flp-HxcPnks5qgOV5gaJpZM4Jhprs>
> .
>
-- 
Richard

outofcoffee commented:
@rnorth #189 solves something different. In this scenario, PWD resolves to a path within the current container; namely the one running the JUnit test. That path is not going to be accessible to the Compose container, so Compose can't see the YAML file.

Mounting a volume is also impractical as a work around due to the issue of the buildup of garbage with containers having different lifecycles (compose and the test host container).

In short, no, it doesn't work for this scenario.

outofcoffee commented:
@rnorth rebased on master and:

* switched to zt-exec
* cleaned up exit value check
* tidied up a bit
* made `withLocalCompose` accept a `boolean`

rnorth commented:
@outofcoffee sorry for the numerous delays in looking at this. I will merge. However, please could you rebase onto master and solve the current merge conflicts?

pcornish commented:
@rnorth  - passed on Travis, but CircleCI failed with:
```
org.testcontainers.containers.ContainerFetchException: Retry limit reached while trying to pull image: mysql:5.5
```

:-/

rnorth commented:
@pcornish thanks. Docker pull issue = :z 
Merging, as this failure is an intermittent one.
